ProcFilter.cpp

```diff
@@ -0,0 +1,355 @@
+ * There are two version of libaudit (libaudit0, and libaudit1) this makes it
+ * impossible to build once then run on all supported distro versions.
+ *
+ * But, since libauparse is available on all supported distros, and it also
+ * links to libaudit, all we need to do is call dlsym to get the function
+ * pointer(s) we need.
+ *
+ *****************************************************************************/
+
+
+
+/*****************************************************************************
+ ** ProcessInfo
+ *****************************************************************************/
+
+ProcessInfo::ProcessInfo(std::string description, int processId, int parentProcessId)
+{
+    name = description;
+    pid = processId;
+    ppid = parentProcessId;
+}
+
+const ProcessInfo ProcessInfo::Empty("0",0,0);
+
+inline bool ProcessInfo::operator==(const ProcessInfo& x) const
+{
+    return (pid == x.pid) && (ppid == x.ppid) && (name == x.name);
+}
+
+inline bool ProcessInfo::operator!=(const ProcessInfo& x) const
+{
+    return !(*this==x);
+}
+
+
+/*****************************************************************************
+ ** ProcFilter
+ *****************************************************************************/
+
+ProcFilter* ProcFilter::_instance = NULL;
+
+std::set<std::string> ProcFilter::_blocked_process_names;
+std::set<std::string> ProcFilter::_blocked_user_names;
+
+// -------- helper functions -----------------------------
+int ProcFilter::is_dir(std::string path)
+{
+    struct stat stat_buf;
+    stat(path.c_str(), &stat_buf);
+    int is_directory = S_ISDIR(stat_buf.st_mode);
+    return (is_directory ? 1: 0);
+}
+
+bool ProcFilter::is_number(const std::string& s)
+{
+    return !s.empty() &&
+           std::find_if(s.begin(), s.end(), [](char c) { return !std::isdigit(c); }) == s.end();
+}
+
+bool ProcFilter::is_process_running(int pid)
+{
+    std::stringstream ss_file;
+    ss_file << "/proc/" << pid << "/stat";
+    std::string file = ss_file.str();
+    struct stat stat_buf;
+    return (stat(file.c_str(), &stat_buf) == 0);
+}
+
+std::string ProcFilter::get_user_of_process(int pid)
+{
+    std::stringstream ss_file;
+    ss_file << "/proc/" << pid;
+    std::string file_name = ss_file.str();
+
+    struct stat stat_buf;
+    if(stat(file_name.c_str(), &stat_buf) == -1)
+    {
+        return std::string();
+    }
+    struct passwd *pw = getpwuid(stat_buf.st_uid);
+    return pw->pw_name;
+}
+
+ProcessInfo ProcFilter::read_proc_data_from_stat(const std::string& fileName)
+{
+    ProcessInfo procData = ProcessInfo::Empty;
+    std::ifstream infile(fileName, std::ifstream::in);
+
+    if(!infile) {
+        return ProcessInfo::Empty;
+    }
+
+    int pid;
+    std::string comm;
+    char state;
+    int ppid;
+    int pgrp;
+    int session;
+    char c;
+
+    if (infile >> pid >> comm >> state >> ppid >> pgrp >> session)
+    {
+        procData.pid = pid;
+        procData.ppid = ppid;
+        procData.name = comm;
```

Comment:
Do not use "comm" it is too easy to bypass. Use /proc/<pid>/exe instead (obtain via readlink()).

Reply:
taking the process name from exe. Also updated the configuration to hold the whole path and comparison changed to "starts with" instead of "contains".